File: _archive_go/internal/service/health_service.go
package service

import (
	"context"
	"math"
	"time"

	"finance-tracker/internal/domain"
	"finance-tracker/internal/repository"
)

// HealthServiceImpl сервис расчёта финансового здоровья
type HealthServiceImpl struct {
	accountRepo        repository.AccountRepository
	txRepo             repository.TransactionRepository
	paymentRepo        repository.PaymentRepository
	categoryBudgetRepo repository.CategoryBudgetRepository
	analyticsRepo      repository.AnalyticsRepository
	goalRepo           repository.GoalRepository
	currencySvc        CurrencyService
	budgetSvc          BudgetService
}

// NewHealthService создаёт новый сервис здоровья
func NewHealthService(
	accountRepo repository.AccountRepository,
	txRepo repository.TransactionRepository,
	paymentRepo repository.PaymentRepository,
	categoryBudgetRepo repository.CategoryBudgetRepository,
	analyticsRepo repository.AnalyticsRepository,
	goalRepo repository.GoalRepository,
	currencySvc CurrencyService,
	budgetSvc BudgetService,
) *HealthServiceImpl {
	return &HealthServiceImpl{
		accountRepo:        accountRepo,
		txRepo:             txRepo,
		paymentRepo:        paymentRepo,
		categoryBudgetRepo: categoryBudgetRepo,
		analyticsRepo:      analyticsRepo,
		goalRepo:           goalRepo,
		currencySvc:        currencySvc,
		budgetSvc:          budgetSvc,
	}
}

// CalculateHealth рассчитывает метрики финансового здоровья
func (s *HealthServiceImpl) CalculateHealth(ctx context.Context) (*domain.FinancialHealth, error) {
	health := &domain.FinancialHealth{}
	now := time.Now()
	currentMonth := now.Format("2006-01")
	prevMonth := now.AddDate(0, -1, 0).Format("2006-01")
	dayOfMonth := now.Day()

	// Получаем баланс
	balance, err := s.accountRepo.GetBalance(ctx)
	if err != nil {
		return nil, err
	}

	// Получаем данные за текущий и прошлый месяц
	currentIncome, currentExpenses, currentSavings, _ := s.analyticsRepo.GetMonthTotals(ctx, currentMonth)
	prevIncome, prevExpenses, prevSavings, _ := s.analyticsRepo.GetMonthTotals(ctx, prevMonth)

	// Получаем данные за 3 месяца для расчёта средних
	month2 := now.AddDate(0, -2, 0).Format("2006-01")
	_, expenses2, _, _ := s.analyticsRepo.GetMonthTotals(ctx, month2)

	// Рассчитываем среднедневные расходы на основе истории (более надёжно)
	// Используем прошлые 2 месяца + текущий пропорционально
	totalHistoricalExpenses := prevExpenses + expenses2
	historicalDays := 60 // ~2 месяца
	
	var avgDailyExpenses float64
	if totalHistoricalExpenses > 0 {
		avgDailyExpenses = totalHistoricalExpenses / float64(historicalDays)
	}
	
	// Если в начале месяца (< 15 дней) — используем прошлый месяц для метрик
	// Иначе используем текущий
	useCurrentMonth := dayOfMonth >= 15

	var referenceIncome, referenceExpenses, referenceSavings float64
	if useCurrentMonth && currentIncome > 0 {
		referenceIncome = currentIncome
		referenceExpenses = currentExpenses
		referenceSavings = currentSavings
	} else {
		// Используем прошлый месяц как базу
		referenceIncome = prevIncome
		referenceExpenses = prevExpenses
		referenceSavings = prevSavings
	}

	// Базовые метрики
	if referenceIncome > 0 {
		health.SavingsRate = (referenceSavings / referenceIncome) * 100
		health.ExpenseToIncome = (referenceExpenses / referenceIncome) * 100
	}

	// Burn rate — среднедневные расходы (на основе истории)
	health.DailySpendingAvg = avgDailyExpenses
	health.BurnRate = avgDailyExpenses

	// На сколько дней хватит баланса (при средних расходах)
	if health.BurnRate > 0 {
		health.DaysUntilZero = int(balance / health.BurnRate)
		health.EmergencyFundDays = health.DaysUntilZero
	}

	// Тренды — сравниваем прошлый месяц с позапрошлым (более стабильно)
	_, prevPrevExpenses, _, _ := s.analyticsRepo.GetMonthTotals(ctx, month2)
	if prevPrevExpenses > 0 {
		health.ExpenseGrowth = ((prevExpenses - prevPrevExpenses) / prevPrevExpenses) * 100
	}
	// Для дохода сравниваем текущий с прошлым только если есть данные
	if prevIncome > 0 && currentIncome > 0 {
		health.IncomeGrowth = ((currentIncome - prevIncome) / prevIncome) * 100
	}
	if prevSavings > 0 && referenceSavings > 0 {
		health.SavingsGrowth = ((referenceSavings - prevSavings) / prevSavings) * 100
	}

	// Бюджеты категорий - считаем превышенные и сохраняем список
	budgets, err := s.categoryBudgetRepo.GetByMonth(ctx, currentMonth)
	if err == nil {
		for _, b := range budgets {
			if b.IsExceeded {
				health.OverBudgetCount++
				overAmount := b.SpentAmount - b.LimitAmount
				overPercent := 0.0
				if b.LimitAmount > 0 {
					overPercent = (overAmount / b.LimitAmount) * 100
				}
				health.OverBudgetList = append(health.OverBudgetList, domain.OverBudgetInfo{
					CategoryName: b.CategoryName,
					BudgetAmount: b.LimitAmount,
					SpentAmount:  b.SpentAmount,
					OverAmount:   overAmount,
					OverPercent:  overPercent,
				})
			}
		}
	}

	// Предстоящие платежи (7 дней)
	nextWeek := now.AddDate(0, 0, 7)
	reminders, totalPayments, _ := s.paymentRepo.GetPaymentsUntilDate(ctx, nextWeek)
	health.UpcomingPayments = totalPayments
	_ = reminders // Используем только totalPayments

	// Payment coverage
	if totalPayments > 0 {
		health.PaymentCoverage = balance / totalPayments
	} else {
		health.PaymentCoverage = 100 // Нет платежей = отличное покрытие
	}

	// Прогноз на конец месяца
	daysLeft := time.Date(now.Year(), now.Month()+1, 0, 0, 0, 0, 0, time.UTC).Day() - now.Day()
	health.PredictedEndOfMonth = balance - (health.BurnRate * float64(daysLeft))

	// Получаем обязательные ежемесячные платежи (аренда, кредиты и т.д.)
	var monthlyPayments float64
	payments, err := s.paymentRepo.GetActive(ctx)
	if err == nil {
		for _, p := range payments {
			// Только обязательные и не разовые
			if p.Category == "essential" && !p.IsOneTime {
				// Конвертируем в BYN если нужно
				amount := p.Amount
				if p.Currency != "BYN" && s.currencySvc != nil {
					rate, err := s.currencySvc.GetRate(ctx, p.Currency)
					if err == nil && rate > 0 {
						amount = p.Amount * rate
					}
				}
				monthlyPayments += amount
			}
		}
	}
	dailyPayments := monthlyPayments / 30.0 // Обязательные платежи в пересчёте на день

	// Копилка (подушка безопасности)
	if s.goalRepo != nil {
		goal, err := s.goalRepo.GetActive(ctx)
		if err == nil && goal != nil {
			health.GoalName = goal.Name
			health.TotalSavingsUSD = goal.CurrentAmount
			health.GoalProgress = 0
			if goal.TargetAmount > 0 {
				health.GoalProgress = (goal.CurrentAmount / goal.TargetAmount) * 100
			}

			// Конвертируем накопления в BYN для расчёта дней
			if s.currencySvc != nil {
				usdRate, err := s.currencySvc.GetRate(ctx, "USD")
				if err == nil && usdRate > 0 {
					health.TotalSavings = goal.CurrentAmount * usdRate
					// На сколько дней хватит копилки
					// Учитываем и повседневные расходы, и обязательные платежи
					totalDailyExpenses := health.BurnRate + dailyPayments
					if totalDailyExpenses > 0 {
						health.SavingsDays = int(health.TotalSavings / totalDailyExpenses)
					}
				}
			}
		}
	}

	// Получаем Cashflow для учёта дефицита
	if s.budgetSvc != nil {
		cashflow, err := s.budgetSvc.CalculateCashflow(ctx)
		if err == nil {
			health.CashflowFree = cashflow.FreeFunds
			health.CashflowDeficit = cashflow.FreeFunds < 0
		}
	}

	// Расчёт Health Score (0-100)
	health.HealthScore = s.calculateHealthScore(health)
	health.Status, health.Message = s.getStatusAndMessage(health)

	return health, nil
}

// calculateHealthScore рассчитывает общий скоринг здоровья
func (s *HealthServiceImpl) calculateHealthScore(h *domain.FinancialHealth) int {
	score := 100.0

	// КРИТИЧНО: Дефицит Cashflow (нет денег до ЗП)
	if h.CashflowDeficit {
		// Чем больше дефицит, тем хуже
		if h.CashflowFree < -500 {
			score -= 50 // Серьёзный дефицит
		} else if h.CashflowFree < -100 {
			score -= 35 // Значительный дефицит
		} else {
			score -= 25 // Небольшой дефицит
		}
	}

	// Savings Rate (20% идеально)
	if h.SavingsRate < 0 {
		score -= 20
	} else if h.SavingsRate < 10 {
		score -= 10
	} else if h.SavingsRate < 20 {
		score -= 5
	}

	// Expense to Income (< 80% хорошо)
	if h.ExpenseToIncome > 100 {
		score -= 25
	} else if h.ExpenseToIncome > 90 {
		score -= 15
	} else if h.ExpenseToIncome > 80 {
		score -= 5
	}

	// Баланс (на сколько дней хватит)
	if h.EmergencyFundDays < 7 {
		score -= 10
	} else if h.EmergencyFundDays < 14 {
		score -= 5
	}

	// Копилка / подушка безопасности (рекомендуется 3-6 месяцев = 90-180 дней)
	if h.SavingsDays <= 0 {
		score -= 20 // Нет подушки — серьёзный риск
	} else if h.SavingsDays < 30 {
		score -= 15 // Меньше месяца — опасно
	} else if h.SavingsDays < 60 {
		score -= 10 // 1-2 месяца — недостаточно
	} else if h.SavingsDays < 90 {
		score -= 5 // 2-3 месяца — нормально
	}
	// >= 90 дней — хорошо, бонус не нужен

	// Превышенные бюджеты
	score -= float64(h.OverBudgetCount * 5)

	// Payment Coverage
	if h.PaymentCoverage < 1 {
		score -= 15
	} else if h.PaymentCoverage < 1.5 {
		score -= 5
	}

	// Expense Growth (рост расходов плохо)
	if h.ExpenseGrowth > 30 {
		score -= 10
	} else if h.ExpenseGrowth > 15 {
		score -= 5
	}

	return int(math.Max(0, math.Min(100, score)))
}

// getStatusAndMessage возвращает статус и сообщение на основе скоринга
func (s *HealthServiceImpl) getStatusAndMessage(h *domain.FinancialHealth) (string, string) {
	// Приоритет: дефицит важнее всего
	if h.CashflowDeficit {
		if h.CashflowFree < -500 {
			return "critical", "🚨 Серьёзный дефицит! Денег не хватит до зарплаты. Срочно сократите расходы."
		}
		return "warning", "⚠️ Дефицит до ЗП! Рекомендуется сократить необязательные расходы."
	}

	switch {
	case h.HealthScore >= 80:
		if h.SavingsDays >= 90 {
			return "excellent", "🌟 Отличное состояние! Подушка безопасности в норме."
		}
		return "excellent", "🌟 Хорошее состояние! Рекомендуем увеличить подушку до 3 месяцев."
	case h.HealthScore >= 60:
		msg := "👍 Хорошее состояние. "
		if h.SavingsDays < 30 {
			msg += "Подушка слишком маленькая! "
		} else if h.SavingsDays < 60 {
			msg += "Увеличьте подушку безопасности. "
		}
		if h.OverBudgetCount > 0 {
			msg += "Следите за бюджетами. "
		}
		return "good", msg
	case h.HealthScore >= 40:
		msg := "⚠️ Внимание! "
		if h.SavingsDays <= 0 {
			msg += "Нет подушки безопасности! "
		} else if h.SavingsDays < 30 {
			msg += "Подушка меньше месяца — рискованно. "
		}
		if h.PaymentCoverage < 1 {
			msg += "Баланс не покрывает платежи. "
		}
		if h.OverBudgetCount > 0 {
			msg += "Превышены бюджеты. "
		}
		return "warning", msg
	default:
		if h.SavingsDays <= 0 {
			return "critical", "🚨 Критично! Нет подушки безопасности. Срочно начните откладывать."
		}
		return "critical", "🚨 Критическое состояние! Требуется пересмотр расходов."
	}
}

File: api/app/Services/System/HealthService.php
<?php

namespace App\Services\System;

use App\DTOs\HealthData;
use App\Models\CategoryBudget;
use App\Models\Goal;
use App\Models\Transaction;
use App\Repositories\TransactionRepositoryInterface;
use App\Services\Accounts\AccountService;
use Illuminate\Support\Facades\DB;

class HealthService
{
    public function __construct(
        protected TransactionRepositoryInterface $transactionRepository,
        protected AccountService $accountService
    ) {}

    public function calculateHealth(int $clientId): HealthData
    {
        $currentMonth = now()->format('Y-m');
        $prevMonth = now()->subMonth()->format('Y-m');
        $prevPrevMonth = now()->subMonths(2)->format('Y-m');

        $balance = $this->accountService->getTotalBalance($clientId);

        $current = $this->getMonthTotals($clientId, $currentMonth);
        $prev = $this->getMonthTotals($clientId, $prevMonth);
        $prevPrev = $this->getMonthTotals($clientId, $prevPrevMonth);

        $usdRate = $this->getUsdRate($clientId);

        $health = new HealthData(predicted_end_of_month: $balance);

        $refIncome = $current['income'] > 0 ? $current['income'] : $prev['income'];
        $refExpenses = $current['expenses'] > 0 ? $current['expenses'] : $prev['expenses'];
        $refSavings = $current['savings'] > 0 ? $current['savings'] : $prev['savings'];

        if ($refIncome > 0) {
            $health->savings_rate = ($refSavings / $refIncome) * 100;
            $health->expense_to_income = ($refExpenses / $refIncome) * 100;
        }

        $totalHistoricalExpenses = $prev['expenses'] + $prevPrev['expenses'];
        $avgDailyExpenses = $totalHistoricalExpenses / 60;
        $health->burn_rate = $avgDailyExpenses;
        $health->daily_spending_avg = $avgDailyExpenses;

        if ($avgDailyExpenses > 0) {
            $health->emergency_fund_days = (int) ($balance / $avgDailyExpenses);
            $health->days_until_zero = $health->emergency_fund_days;
        }

        if ($prevPrev['expenses'] > 0) {
            $health->expense_growth = (($prev['expenses'] - $prevPrev['expenses']) / $prevPrev['expenses']) * 100;
        }
        if ($prev['income'] > 0 && $current['income'] > 0) {
            $health->income_growth = (($current['income'] - $prev['income']) / $prev['income']) * 100;
        }
        if ($prev['savings'] > 0 && $refSavings > 0) {
            $health->savings_growth = (($refSavings - $prev['savings']) / $prev['savings']) * 100;
        }

        $goals = Goal::withoutGlobalScope('client')->where('client_id', $clientId)->where('is_active', true)->get();
        $activeGoalIds = $goals->pluck('id')->all();
        $totalSavings = $this->transactionRepository->getTotalSavings($clientId);
        $cushionSavings = $this->transactionRepository->getTotalSavingsActiveGoalsOnly($clientId, $activeGoalIds);

        $health->total_savings = $totalSavings;
        $health->total_savings_usd = $usdRate > 0 ? $totalSavings / $usdRate : 0;
        if ($avgDailyExpenses > 0) {
            $health->savings_days = (int) ($cushionSavings / $avgDailyExpenses);
        }

        if ($goals->isNotEmpty()) {
            $rates = $this->getRates($clientId);
            $totalTargetBYN = 0;
            foreach ($goals as $g) {
                $target = (float) $g->target_amount;
                $currency = $g->currency ?? 'BYN';
                $targetBYN = $this->convertToBYN($target, $currency, $rates);
                $totalTargetBYN += $targetBYN;
            }
            $health->goal_name = $goals->count() > 1
                ? $goals->pluck('name')->implode(', ')
                : $goals->first()->name;
            if ($totalTargetBYN > 0) {
                $health->goal_progress = ($health->total_savings / $totalTargetBYN) * 100;
            }
            if ($goals->count() === 1) {
                $firstGoal = $goals->first();
                $savedForGoal = $this->transactionRepository->getTotalSavingsForGoal($clientId, $firstGoal->id);
                $currency = $firstGoal->currency ?? 'BYN';
                $savedInGoalCurrency = $currency === 'BYN'
                    ? $savedForGoal
                    : round($savedForGoal / ($rates[$currency] ?? 1), 2);
                $target = (float) $firstGoal->target_amount;
                $health->first_goal_savings = $savedForGoal;
                $health->first_goal_savings_usd = $usdRate > 0 ? $savedForGoal / $usdRate : 0;
                $health->first_goal_progress = $target > 0 ? ($savedInGoalCurrency / $target) * 100 : 0;
            }
        }

        $overBudget = $this->getOverBudgetCategories($clientId, $currentMonth);
        $health->over_budget_count = count($overBudget);
        $health->over_budget_list = $overBudget;

        $daysLeft = (int) now()->diffInDays(now()->copy()->endOfMonth(), false) + 1;
        $health->predicted_end_of_month = $balance - ($avgDailyExpenses * $daysLeft);

        $totalDebt = (float) DB::table('debts')
            ->where('client_id', $clientId)
            ->sum(DB::raw('total_amount - paid_amount'));
        $health->total_debt = $totalDebt;

        if ($refIncome > 0) {
            $health->debt_to_income = ($totalDebt / $refIncome) * 100;
        }

        $health->net_worth = $balance + $health->total_savings - $totalDebt;

        [$health->health_score, $health->status, $health->message] = $this->calculateScore($health);

        return $health;
    }

    protected function getMonthTotals(int $clientId, string $month): array
    {
        $income = $this->transactionRepository->getIncomeForMonth($clientId, $month);
        $expenses = $this->transactionRepository->getExpensesForMonth($clientId, $month);
        $savings = $this->transactionRepository->getSavingsForMonth($clientId, $month);

        return ['income' => $income, 'expenses' => $expenses, 'savings' => $savings];
    }

    protected function getUsdRate(int $clientId): float
    {
        return $this->getRates($clientId)['USD'];
    }

    protected function getRates(int $clientId): array
    {
        $rows = DB::table('settings')
            ->where('client_id', $clientId)
            ->whereIn('key', ['usd_rate', 'eur_rate', 'rub_rate'])
            ->get()
            ->keyBy('key');

        return [
            'BYN' => 1,
            'USD' => $rows->get('usd_rate') ? (float) $rows->get('usd_rate')->value : 3.25,
            'EUR' => $rows->get('eur_rate') ? (float) $rows->get('eur_rate')->value : 3.55,
            'RUB' => $rows->get('rub_rate') ? (float) $rows->get('rub_rate')->value : 0.034,
        ];
    }

    protected function convertToBYN(float $amount, string $currency, array $rates): float
    {
        if ($currency === 'BYN') {
            return $amount;
        }
        $rate = $rates[$currency] ?? 1;

        return $rate > 0 ? $amount * $rate : 0;
    }

    protected function getOverBudgetCategories(int $clientId, string $month): array
    {
        $budgets = CategoryBudget::withoutGlobalScope('client')
            ->with('category')
            ->where('client_id', $clientId)
            ->where('month', $month)
            ->get();

        $result = [];
        foreach ($budgets as $b) {
            $spent = (float) Transaction::withoutGlobalScope('client')
                ->where('client_id', $clientId)
                ->where('month', $month)
                ->where('category_id', $b->category_id)
                ->where('type', 'expense')
                ->sum(DB::raw('ABS(amount)'));
            if ($b->limit_amount > 0 && $spent > $b->limit_amount) {
                $overAmount = $spent - $b->limit_amount;
                $overPercent = ($overAmount / $b->limit_amount) * 100;
                $result[] = [
                    'category_name' => $b->category->name ?? 'Категория',
                    'budget_amount' => (float) $b->limit_amount,
                    'spent_amount' => $spent,
                    'over_amount' => $overAmount,
                    'over_percent' => $overPercent,
                ];
            }
        }

        return $result;
    }

    protected function calculateScore(HealthData $h): array
    {
        $score = 50;
        if ($h->savings_rate >= 20) {
            $score += 15;
        } elseif ($h->savings_rate >= 10) {
            $score += 5;
        } elseif ($h->savings_rate >= 0) {
            $score -= 5;
        } else {
            $score -= 15;
        }

        if ($h->expense_to_income <= 70) {
            $score += 10;
        } elseif ($h->expense_to_income <= 85) {
            $score += 5;
        } elseif ($h->expense_to_income <= 95) {
            $score -= 5;
        } else {
            $score -= 15;
        }

        if ($h->emergency_fund_days >= 90) {
            $score += 15;
        } elseif ($h->emergency_fund_days >= 30) {
            $score += 5;
        } elseif ($h->emergency_fund_days >= 14) {
            $score -= 5;
        } else {
            $score -= 15;
        }

        if ($h->expense_growth <= 0) {
            $score += 5;
        } elseif ($h->expense_growth <= 15) {
            $score += 0;
        } else {
            $score -= 10;
        }

        if ($h->goal_progress >= 100) {
            $score += 10;
        }

        if ($h->debt_to_income >= 50) {
            $score -= 20;
        } elseif ($h->debt_to_income >= 30) {
            $score -= 10;
        }

        if ($h->net_worth > 0) {
            $score += 5;
        }

        $score = max(0, min(100, $score));

        $message = $h->goal_progress >= 100 ? ' Цель накоплений достигнута!' : '';

        if ($score >= 80) {
            return [$score, 'excellent', 'Финансовое здоровье отличное.'.$message];
        }
        if ($score >= 60) {
            return [$score, 'good', 'Финансовое здоровье хорошее.'.$message];
        }
        if ($score >= 40) {
            return [$score, 'warning', 'Есть возможности для улучшения.'.$message];
        }

        return [$score, 'critical', 'Требуется внимание к финансам.'.$message];
    }
}
